### PR TITLE
Influxql bitwise operators

### PR DESCRIPTION
**Pull request by Tomcat-Engineering**
Relates to #8042 

This implements bitwise AND, OR and XOR operators.   They can be used in both the main part of a SELECT statement, and in a WHERE clause.  They accept any non-negative integer (field or constant) on either side.

There is no bitwise-NOT operator, for reasons explained in [the documentation](https://github.com/Tomcat-Engineering/docs.influxdata.com/blob/add-bitwise-operators/content/influxdb/v1.2/query_language/math_operators.md#bitwise-not).  There is a simple workaround which avoids the need for a new influxql function, also explained in the documentation.

Negative numbers are not supported, because representing a bitfield as a negative number is a very advanced use-case, involving detailed knowledge of twos-complement arithmetic and the underlying `int64` implementation.  In practice all this means is that we support bitfields up to 63 bits wide instead of 64.  If you try to use one of these operators with a negative number then you just get 0 as a result - this is the same behaviour as InfluxQL uses for other "invalid operands" like divide-by-zero errors.

###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [x] [InfluxQL Spec](https://github.com/influxdata/influxdb/blob/master/influxql/README.md) updated
- [x] [InfluxData Documentation](https://github.com/influxdata/docs.influxdata.com): https://github.com/influxdata/docs.influxdata.com/pull/1067

### CODE REVIEW

**Comment by jsternberg** [comment]
What's the rationale for restricting this to positive numbers? It might be worth limiting this to int64's only rather than floats, but I'm not sure why any of the relevant operators are affected by if they are positive or negative.

**Comment by jwilder** [comment]
Should these support `boolean` fields as well?

**Comment by Tomcat-Engineering** [comment]
@jsternberg The rationale for excluding negative integers was explained in the description of the pull request and on the ticket (#8042),  but it seems to be confusing people so I have removed the restriction.

The new operators don't work with floats.

@jwilder I am adding support for booleans too.  But you won't be able to mix integers and booleans.

File: influxql/select.go [hunk 1]
```diff
@@ -1429,6 +1503,16 @@ func buildTransformIterator(lhs Iterator, rhs Iterator, op Token, opt IteratorOp
...
 		}
 		return newIntegerBooleanExprIterator(left, right, opt, fn), nil
+	case func(bool, bool) bool:
+		left, ok := lhs.(BooleanIterator)
+		if !ok {
+			return nil, fmt.Errorf("type mismatch on LHS, unable to use %T as a BooleanIterator", lhs)
+		}
+		right, ok := rhs.(BooleanIterator)
+		if !ok {
+			return nil, fmt.Errorf("type mismatch on LHS, unable to use %T as a BooleanIterator", rhs)
```
Comment (jsternberg) [inline influxql/select.go]: This looks like a copy/paste error. Shouldn't this be `RHS`?

File: influxql/select_test.go [hunk 2]
```diff
@@ -2284,6 +2284,33 @@ func TestSelect_BinaryExpr_Integer(t *testing.T) {
 				{&influxql.FloatPoint{Name: "cpu", Time: 9 * Second, Value: 1}},
 			},
 		},
+		{
```
Comment (jsternberg) [inline influxql/select_test.go]: Can we also get tests added for the boolean points?

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 ### Features
 
+- [#8042](https://github.com/influxdata/influxdb/issues/8042): Add bitwise AND, OR and XOR operators to the query language
```
Comment (jsternberg) [inline CHANGELOG.md]: Minor nitpick. Can you move this to the bottom after you rebase? This is really annoying me for some reason, but it really shouldn't annoy me.

**Comment by jsternberg** [comment]
After the unit tests pass, can you squash this to a single commit and then we'll merge it? Thanks. It looks good to me at the moment.

**Comment by jsternberg** [comment]
Can you squash the changelog tweak into the first commit?

**Comment by Tomcat-Engineering** [comment]
Done - sorry, thought you had a magic "squash and merge" button these days!

**Comment by jsternberg** [comment]
Apparently, I do have a button that does that. I was not aware of that. Thanks though! I guess I'll try that out next time.

**Pull request merged**